text/utilities/0000-add-live-reload-library.md

```diff
@@ -0,0 +1,69 @@
+# Add Process Reload library to Paketo
+
+## Summary
+
+Add a standalone library for buildpack authors to quickly incorporate process reload into their buildpacks.
+
+## Motivation
+
+Currently, implementing live reload in a buildpack requires the buildpack author to have knowledge of watchexec internals,
+such as command line parameters, and have to know when to enable live reload (typically `BP_LIVE_RELOAD_ENABLED`).
+
+The typical implementation pattern involves finding an example buildpack and copy/pasting implementation code from that buildpack
+into the new buildpack. This RFC proposes a helper library to implement live reload with a clean division of interface
+("live reload") and implementation ("watchexec").
+
+## Detailed Explanation
+
+Create a new repository to host the library: `paketo-buildpacks/reload`. See [Implementation](#implementation) for further details.
+
+Going forward, buildpack authors can rely on this library to assist with reloadable processess, instead of making large 
+PRs that require intimate knowledge of the `watchexec` CLI. Examples:
+
+- [npm-start](https://github.com/paketo-buildpacks/npm-start/pull/160)
+- [python-start](https://github.com/paketo-buildpacks/python-start/pull/79)
+- [go-build](https://github.com/paketo-buildpacks/go-build/pull/237)
+
+## Rationale and Alternatives
+
+1. Include the library code in packit. See [PR 343](https://github.com/paketo-buildpacks/packit/pull/343). 
+While we did consider the possibility of including the interface ("live reload") in Packit, we knew that the implementation
+("watchexec") was too tightly coupled with `watchexec` to live in Packit.
+2. Include the library code in [the `watchexec` buildpack](https://github.com/paketo-buildpacks/watchexec).
+We felt that we should not include both library code and buildpack code in the buildpack repository, especially since
+in this case having a generic interface ("live reload") implies that eventually another implementation might be possible.
+
+## Implementation
+
+Create a new repository to host the library: `paketo-buildpacks/reload`. The root package `reload` will contain an interface
+that looks something like this:
+
+```go
+package reload
+
+import packit "github.com/paketo-buildpacks/packit/v2"
```

Comment:
A comment/challenge, could we make this work in a library-agnostic way? Without sacrificing too much on the API.

Alternatively, could we hide the library-specific bits in their own packages so we could support both packit and libcnb without needing to include both projects in a consuming buildpack?

Reply:
I was hoping that conversion might be an option, but the structs in packit and libcnb are slightly different and Go doesn't like that, so it won't convert them.

I haven't played with generics yet as it's new in 1.18 and we haven't updated to that yet on the Java buildpacks side. I would not be opposed to it though, we'll be updating to 1.18 soon enough.

Anyway, my concern is that if we add a library for only one of the two libraries being used by Paketo buildpacks, we're making things harder on buildpack authors. It's already tough enough given we have the library split, if we can make it work with both that would prevent some friction. Hopefully, in the not too distant future, we'll have a packit that's based on libcnb and we'll be sharing the same core structs.

Reply:
I've implemented a beta over at https://github.com/joshuatcasey/libreload-beta but it's still very coupled to `packit v2`. I'll see if I can work out a way to de-couple the `packit` dependency.

Reply:
>How close are we to having a libcnb v2?

I can't say exactly and I think it depends somewhat on how much effort there is to implement what still needs to be done for a v2. 

What I can say is that we've switched the main branch of libcnb to v2, which I think is a good step/sign that we're getting close. I will ask around more and see if I can get a feel for how much work is left.

Reply:
> I don't think generics will help, because you'll still have to import both libraries if you want concrete instantiations of each library's type.

Yeah, this was the point I was trying to make above. I think for now we should just pick `packit v2` and try that - there are about eight packit-based buildpacks that could benefit from this utility, so we get a chance to standardize many different implementations of `watchexec`.

There will also be some lag between `libcnb v2` release and being able to incorporate that into the existing buildpacks that we have. I think we could have a `libreload v1` based on `packit v2` and a `libreload v2` based on `libcnb v2`.

Reply:
Why not put this into packit then? It'll be tied to packit. If that's not an option, I know it was discussed, I think we should perhaps call the proposed library `packit-reload` or something to indicate it's based on packit.

I would like to see us targeting libcnb v2, but I guess it depends on the speed at which we need a solution here. We discussed in the BAT team meeting today and I'd say we're in the milestone stage for a v2 release. We'd like people to start using it and provide some feedback, at the same time, things could still change before we release it.

>I think we could have a libreload v1 based on packit v2 and a libreload v2 based on libcnb v2.

I worry this would be confusing to users, I think that's why a name like `packit-reload` would be more suitable. It's clear what you get.

I worry changing versions could also create a maintenance burden. If there's a v1 release based on packit now, and in three months we cut a v2, how long do we continue to support v1?

Reply:
I'm 👍 for that name. 

I'm not sure that having a separate repo addresses the concerns about maintenance burden, but I'll defer to @ryanmoran  on that since this is now a packit specific library.

Reply:
I think creating a new repo called `libreload-packit` is fine. I think my objection to including this code in `packit` is that it is highly coupled to a specific set of buildpacks while the rest of `packit` is generic to any type of buildpack a developer might want to make.